### ISSUE TITLE
CVE needed for the recently fixed signed integer overflow in SafeInt

### ISSUE THREAD

**Issue by practicalswift** [dcleblanc/SafeInt#11]
First, thanks a lot for the work on this library - I appreciate it and the fact that it was released by Microsoft under a permissive license.

Second, thanks for fixing the reported issue quickly. That is much appreciated too!

Now that the fix is public I can post this in the ordinary issue tracker. This will allow others to chime in too.

**Background**

On November 14, 2019 I submitted a bug report privately regarding a signed integer overflow I identified in SafeInt while auditing safe integer libraries for potential use in the Bitcoin Core project.

The bug report was sent to Microsoft Security Response Center (MSRC, id: VULN-012305) and to the project maintainer.

The fix was committed to `master` November 15, 2019. That is one day after it was reported: such great turnaround times are very rare. Thanks!

This was the proof of concept I sent in the bug report:

```
$ git clone https://github.com/dcleblanc/SafeInt
$ cd SafeInt
$ git rev-parse HEAD
67e5e5932df3b15239154035fddee4e569c97695
$ cat > safeint-poc.cpp
#include "SafeInt.hpp"

#include <cstdint>

int main() {
  uint16_t i = 46341, j = 46341, k = 0;
  SafeMultiply(i, j, k);
}
$ g++-8 -ftrapv -o safeint-poc safeint-poc.cpp && ./safeint-poc
Aborted (core dumped)
$ clang++-8 -ftrapv -o safeint-poc safeint-poc.cpp && ./safeint-poc
Illegal instruction (core dumped)
$ clang++-8 -fsanitize=undefined -o safeint-poc safeint-poc.cpp && ./safeint-poc
SafeInt.hpp:1836:45: runtime error: signed integer overflow: 46341 * 46341 cannot be represented in type 'int'
```

Documentation for `SafeMultiply`:

> **SafeMultiply**:
> Multiplies two numbers together in a way that protects against overflow.
> 
> ```
> template<typename T, typename U>
> inline bool SafeMultiply (T t, U u, T& result) throw ();
> ```

**CVE request**

As a user of the SafeInt library I would like an issue like this to get assigned a CVE.

A CVE assignment allows me as a user to more easily make an informed decision about needing to upgrade to the fixed version or not.

As I understand it Microsoft is an assigning CNA (CVE Numbering Authority): can MSRC assign a CVE for this issue?

Thanks!

**Comment by dcleblanc** [comment]
I’d like to better understand – you’d like a CVE entry, but I’m not seeing an actual vulnerability. The library behaves as designed, and correctly rejects the inputs on both Linux and Windows. That it had a dependency on non-standard behavior was a flaw, but I’m not aware of a platform where it could have come to the wrong outcome. Theoretically, there could be a platform that did something different with an IMUL instruction, but I’m not aware of any.

The next problem is clearly demonstrated – with this combination of diagnostic flags, the code might crash. But I’m not sure if anyone would reasonably run in production with these flags – it must have some perf impact, and I doubt much code would survive. I’m happy the flags found the issue, and I updated the test harness to ensure that problem won’t happen again. Whether MSRC wants to issue a CVE is up to them – I’m not part of that decision.

To answer the main question – I wouldn’t view taking this fix as an emergency for most uses. Only if you run with those flags set in production would this really be an important fix. If you’re not using those flags, the code still does the right thing.

Thanks for finding it, and now that I know about these flags, I’m glad we’ve further improved the quality of the code. I was also pleasantly surprised that after running the test harness, this was the only instance of that problem found.

**Comment by practicalswift** [comment]
> To answer the main question – I wouldn’t view taking this fix as an emergency for most uses. Only if you run with those flags set in production would this really be an important fix. If you’re not using those flags, the code still does the right thing.

CVE:s are issued for all types of vulnerabilities -- everything from XSS to RCE. Thus "emergency for most users" is clearly not a precondition for a CVE assignment.

We should be conservative and assume that users of a SafeInt are _highly_ sensitive to signed integer overflows since the goal of the library is to provide safe integer operations.

I know of high-frequency trading applications where program termination via a trap is a much better outcome than continued execution in the case of a signed integer overflow.

Running with `-fsanitize-trap=signed-integer-overflow`, `-ftrapv`, or similar makes sense for such an application (since it is reasonably assumed that SafeInt will provide safe integer guarantees).

For such a user this bug is clearly a vulnerability.

(Additionally there are users who would consider _any_ UB in a security critical library such as SafeInt to be a vulnerability.)

Questions:

1. Is there a scenario where users of the library would be worse off by knowing that they might want to consider upgrading if they are sensitive to signed integer overflows for some reason? A CVE assignment increases the likelihood of that information reaching users of the library.

2. If a signed integer overflow is not considered a vulnerability in a safe integer library: what would be? :)

---

Aside:

Some excellent background reading for lurkers in this thread who might be new to the concept of UB in the context of modern optimising compilers:
* [What Every C Programmer Should Know About Undefined Behavior, part 1-3](http://blog.llvm.org/2011/05/what-every-c-programmer-should-know.html)
* [A Guide to Undefined Behavior in C and C++, part 1-3](https://blog.regehr.org/archives/213)

**Comment by dcleblanc** [comment]
I just authored and maintain the code - not my decision to issue CVEs on behalf of my employer. So whether it deserves a CVE is really something to take up with them - they typically only do so when a flaw results in a bulletin. On a 1-5 scale, where 1 is definitely not, and 5 is definitely yes, I'm about a 2-3 on this in my personal opinion.

I wasn't previously aware of this combination of flags, and prior to this check-in, wouldn't consider that a supported configuration. It's an improvement to be able to support that, and prove there's no other undefined behavior (unless there's test coverage missing somewhere, hope not, but if so, I'll fix it).

What I consider a vulnerability is if the code comes to the wrong conclusion - something that's unsafe is judged to be safe (or the other way around). For example, if you look several revisions back, you'll find a bug in the code that deals with initializing the class from a double - that's a vulnerability, though it is very much a corner case, where this is not.

I am a bit cranky about the hard line that some compilers have taken with respect to undefined behavior once we consider that the compiler does know what assembly is being emitted, and what that assembly does. For example, we know that there are not two different instructions for signed and unsigned addition on an Intel-compatible processor (or ARM), and that the bit fields emitted are identical, assuming 2's complement signed numbers (which will soon be the standard, and which this library requires with a static_assert). It might be undefined with respect to the standard, but we know _exactly_ what the assembly is going to do. So doing things like throwing out conditional statements really only serves to create hard to diagnose bugs that can only be found by stepping through release code in assembly, and I don't agree with it.

Some years ago, John Regehr's students did a project on this, and found a few problems, and since then, I've attempted to avoid non-standard behavior. Sadly, I missed one, and I thank you for finding it, and I'm very pleased to be able to enhance my test matrix to include this.

Further, there's a body of undefined behavior still lurking out there, but fortunately, the compilers have all chosen not to torment us with it (yet), which is that casting between signed and unsigned types is also not defined. There's lots of this in the library, and it is one of the things I did to manage the undefined behavior because:

int test = (int)((unsigned)x + (unsigned)y);

is defined with respect to the addition (assuming x,y signed), but at assembly level, it is really the same thing as:

int test = x + y;

Which is undefined in the case of overflow. So if they start getting picky about that, then I'm not sure what I'll do, but will manage something.

As to user cases, yes, I would assume that users of SafeInt would be more sensitive to undefined behaviors, but my assumption is that their primary use case are just overflows. In a large body of code, overflows that could cause a vuln are much more common than undefined behaviors, though both clearly exist, and this also depends on the compiler.

To sum up - if you want to work with CVE to make a CVE to get people to upgrade, I'm always happy for people to stay current as I do keep improving it here and there. However, I don't get to tell MSRC how to do their job.

If you find any other issues, including missing test cases, please do let me know, and I'll fix them as soon as I'm able (depending on priority).

Thanks again for bringing this to my attention - it's only about the 5th actual bug to be found that affects runtime behavior in the 16 years since I first wrote it.
